README.md
# Colored Cat

[![Build](https://github.com/helviojunior/pyccat/actions/workflows/build_and_publish.yml/badge.svg)](https://github.com/helviojunior/pyccat/actions/workflows/build_and_publish.yml)
[![Build](https://github.com/helviojunior/pyccat/actions/workflows/build_and_test.yml/badge.svg)](https://github.com/helviojunior/pyccat/actions/workflows/build_and_test.yml)
[![Downloads](https://pepy.tech/badge/pyccat/month)](https://pepy.tech/project/pyccat)
[![Supported Versions](https://img.shields.io/pypi/pyversions/pyccat.svg)](https://pypi.org/project/pyccat)
[![Contributors](https://img.shields.io/github/contributors/helviojunior/pyccat.svg)](https://github.com/helviojunior/pyccat/graphs/contributors)
[![PyPI version](https://img.shields.io/pypi/v/pyccat.svg)](https://pypi.org/project/pyccat/)
[![License: GPL-3.0](https://img.shields.io/pypi/l/pyccat.svg)](https://github.com/helviojunior/pyccat/blob/main/LICENSE)

CCat officially supports Python 3.8+.

## Main features

* [x] Read and highlight text and code files
* [x] Filter to display only selected lines
* [x] Multiple highlight styles
* [x] Save output at image (png/jpg) file using ansi2image lib.

## Installation

```bash
pip3 install --upgrade pyccat
```

## Help

```bash
ccat -h

positional arguments:
  [filename]                                Filename

Options:
  -s, --simple                              just colorize the file content
  -nt, --no-tabulated                       do not show tab
  --style [style name]                      pygments lib style name. (default: gruvbox-dark). See more at: https://pygments.org/styles/
  -l [filter], --lines [filter]             return only selected lines (ex1: 5:13 or ex2: 50: or ex3: :100)
  -hl [filter], --highlight-lines [filter]  highlight only selected lines (ex1: 5:13 or ex2: 50: or ex3: :100)
  --output-img [filename]                   image output file.
  -h, --help                                show help message and exit
  -v                                        Specify verbosity level (default: 0). Example: -v, -vv, -vvv
  --version                                 show current version
```

## Executing

**Regular linux cat**
![cat](images/regular_cat.jpg)

**Read a file**
```bash
ccat /tmp/teste.json
```

![Sample 001](images/sample_001.jpg)

**Read a file without table**
```bash
ccat -nt /tmp/teste.json
```

![Sample 002](images/sample_002.jpg)

**Just highlight the file**
```bash
ccat -s /tmp/teste.json
```

![Sample 003](images/sample_003.jpg)

**Display only some lines**
```bash
ccat -l 18:37 teste.json
```

![Sample 004](images/sample_004.jpg)

```bash
ccat -l 18:23,35:37 teste.json
```

![Sample 004](images/sample_005.jpg)


**Display only some lines and highlight specific lines**
```bash
ccat -l 18:37 -hl 18:23,35:37 teste.json
```

![Sample 004](images/sample_006.jpg)


CHANGELOG
- 0.1.9: Future version.

- 0.1.8: Save output at image (png/jpg) file using ansi2image lib.
  Bugfix at ``--version`` tag and -h colors.
- 0.1.7: Highlight function with parameter ``-hl``
  Bugfix at old version of library tabulated.
  Update documentation.
- 0.1.6: Table lay-out improvements
  Max width implementation.
- 0.1.5: Bugfix and improvements at cut function
  Max width implementation.
- 0.1.3: Bugfix and improvements at cut function
  Update documentation.
- 0.1.3: Cut function with parameter ``-l``
- 0.1.2: Update documentation.
  Check file size
  Utilization of specific JSON lexer
- 0.1.0: Initial CCat PyPI release.


ccat/args.py
#!/usr/bin/python3
# -*- coding: UTF-8 -*-
from argparse import _ArgumentGroup, Namespace
from .util.color import Color

import argparse, sys, os


class Arguments(object):
    ''' Holds arguments used by the ccat '''
    modules = {}
    verbose = False
    args = None

    def __init__(self):
        self.verbose = any(['-v' in word for word in sys.argv])
        self.args = self.get_arguments()

    def _verbose(self, msg):
        if self.verbose:
            Color.pl(msg)

    def get_arguments(self) -> Namespace:
        ''' Returns parser.args() containing all program arguments '''

        parser = argparse.ArgumentParser(
            usage=argparse.SUPPRESS,
            prog="ccat",
            add_help=False,
            formatter_class=lambda prog: argparse.HelpFormatter(prog, max_help_position=80, width=180))

        parser.add_argument('filename',
                            action='store',
                            metavar='[filename]',
                            type=str,
                            help=Color.s('Filename'))

        flags = parser.add_argument_group('Options')
        self._add_flags_args(flags)

        return parser.parse_args()

    def _add_flags_args(self, flags: _ArgumentGroup):
        flags.add_argument('-s', '--simple',
                           action='store_true',
                           default=False,
                           dest=f'simple',
                           help=Color.s('just colorize the file content'))

        flags.add_argument('-nt', '--no-tabulated',
                           action='store_true',
                           default=False,
                           dest=f'no_tab',
                           help=Color.s('do not show tab'))

        flags.add_argument('--style',
                           action='store',
                           metavar='[style name]',
                           type=str,
                           default='gruvbox-dark',
                           dest=f'style',
                           help=Color.s('pygments lib style name. (default: {G}gruvbox-dark{W}). See more at: https://pygments.org/styles/'))

        flags.add_argument('-l', '--lines',
                           action='store',
                           metavar='[filter]',
                           type=str,
                           default='',
                           dest=f'line_filter',
                           help=Color.s('return only selected lines ({W}{D}ex1:{G} 5:13 {W}{D}or ex2: {G}50: {W}{D}or ex3: {G}:100{W})'))

        flags.add_argument('-hl', '--highlight-lines',
                           action='store',
                           metavar='[filter]',
                           type=str,
                           default='',
                           dest=f'highlight_line_filter',
                           help=Color.s('highlight only selected lines ({W}{D}ex1:{W}{G} 5:13 {W}{D}or ex2: {W}{G}50: {W}{D}or ex3: {W}{G}:100{W})'))

        flags.add_argument('--output-img',
                           action='store',
                           metavar='[filename]',
                           type=str,
                           dest=f'out_file',
                           help=Color.s('image output file.'))

        flags.add_argument('-h', '--help',
                           action='help',
                           help=Color.s('show help message and exit'))

        flags.add_argument('-v',
                           action='count',
                           default=0,
                           help=Color.s(
                               'Specify verbosity level (default: {G}0{W}). Example: {G}-v{W}, {G}-vv{W}, {G}-vvv{W}'
                           ))

        flags.add_argument('--version',
                           action='store_true',
                           default=False,
                           dest=f'version',
                           help=Color.s('show current version'))
